lib/modules/versioning/pvp/index.ts

```diff
@@ -0,0 +1,297 @@
+import { logger } from '../../../logger';
+import type { RangeStrategy } from '../../../types/versioning';
+import { regEx } from '../../../util/regex';
+import type { NewValueConfig, VersioningApi } from '../types';
+
+export const id = 'pvp';
+export const displayName = 'Package Versioning Policy (Haskell)';
+export const urls = [];
+export const supportsRanges = true;
+export const supportedRangeStrategies: RangeStrategy[] = ['auto'];
+
+type Parsed = { lower: string; upper: string };
+type Components = { major: number[]; minor: number[]; patch: number[] };
+
+export function parse(input: string): Parsed | null {
+  const noSpaces = input.replaceAll(' ', '');
+  const r = regEx(/>=(?<lower>[\d.]+)&&<(?<upper>[\d.]+)/);
+  const m = r.exec(noSpaces);
+  if (!m?.groups) {
+    return null;
+  }
+  return {
+    lower: m.groups['lower'],
+    upper: m.groups['upper'],
+  };
+}
+
+export function extractAllComponents(version: string): number[] {
+  const versionMajor = version.split('.');
+  const versionIntMajor: number[] = versionMajor.map((x) => parseInt(x, 10));
+  const ret = [];
+  for (const l of versionIntMajor) {
+    if (l < 0 || !isFinite(l)) {
+      continue;
+    }
+    ret.push(l);
+  }
+  return ret;
+}
+
+function compareIntArray(
+  versionIntMajor: number[],
+  otherIntMajor: number[],
+): 'lt' | 'eq' | 'gt' {
+  for (
+    let i = 0;
+    i < Math.min(versionIntMajor.length, otherIntMajor.length);
+    i++
+  ) {
+    if (versionIntMajor[i] > otherIntMajor[i]) {
+      return 'gt';
+    }
+    if (versionIntMajor[i] < otherIntMajor[i]) {
+      return 'lt';
+    }
+  }
+  if (versionIntMajor.length === otherIntMajor.length) {
+    return 'eq';
+  }
+  if (versionIntMajor.length > otherIntMajor.length) {
```

Comment:
So `1.0 > 1`?

Reply:
Yes, and that's on purpose.

I think equality on a range expression should be defined as it's syntax tree matching (e.g. operators and versions matching). So, if we considered versions `1` and `1.0` equal, ranges `>=1.0 && <1.1` and `>=1 && <1.1` should also be equal. But we can verify that Cabal does regard a version with an appended `.0` as distinct. For example:

```
$ cabal repl --constraint 'base >=4.18.2.1.0'  --with-ghc ghc-9.6.5
[__1] rejecting: base-4.18.2.1/installed-4.18.2.1 (constraint from command line flag requires >=4.18.2.1.0)
$ cabal repl --constraint 'base >=4.18.2.1'  --with-ghc ghc-9.6.5
<gives prompt as expected>